commit f0e0da1ed876ee1becf7950527b6b562e85741c8
Author: Miroslav Rezanina <mrezanin@redhat.com>
Date:   Mon Mar 4 09:05:02 2013 +0100

    char: Split out tcp socket close code in a separate function
    
    Message-Id: <9f0944a25bc1094fa7a74ac9df14e184e2c5c82d.1354903384.git.crobinso@redhat.com>
    
    Signed-off-by: Amit Shah <amit.shah@redhat.com>
    Signed-off-by: Cole Robinson <crobinso@redhat.com>

diff --git a/qemu-char.c b/qemu-char.c
index e4b0f53..d7fa7e6 100644
--- a/qemu-char.c
+++ b/qemu-char.c
@@ -2155,6 +2155,21 @@ typedef struct {
 
 static void tcp_chr_accept(void *opaque);
 
+static void tcp_closed(void *opaque)
+{
+    CharDriverState *chr = opaque;
+    TCPCharDriver *s = chr->opaque;
+
+    s->connected = 0;
+    if (s->listen_fd >= 0) {
+        qemu_set_fd_handler2(s->listen_fd, NULL, tcp_chr_accept, NULL, chr);
+    }
+    qemu_set_fd_handler2(s->fd, NULL, NULL, NULL, NULL);
+    closesocket(s->fd);
+    s->fd = -1;
+    qemu_chr_be_event(chr, CHR_EVENT_CLOSED);
+}
+
 static int tcp_chr_write(CharDriverState *chr, const uint8_t *buf, int len)
 {
     TCPCharDriver *s = chr->opaque;
@@ -2313,15 +2328,7 @@ static void tcp_chr_read(void *opaque)
         len = s->max_size;
     size = tcp_chr_recv(chr, (void *)buf, len);
     if (size == 0) {
-        /* connection closed */
-        s->connected = 0;
-        if (s->listen_fd >= 0) {
-            qemu_set_fd_handler2(s->listen_fd, NULL, tcp_chr_accept, NULL, chr);
-        }
-        qemu_set_fd_handler2(s->fd, NULL, NULL, NULL, NULL);
-        closesocket(s->fd);
-        s->fd = -1;
-        qemu_chr_be_event(chr, CHR_EVENT_CLOSED);
+        tcp_closed(chr);
     } else if (size > 0) {
         if (s->do_telnetopt)
             tcp_chr_process_IAC_bytes(chr, s, buf, &size);
